# Projecting the rate matrix. Prove yourself that the commutation relation works!

PCCA+ is a clustering algorithm. In this work, it is used for computing the transition matrix $K^c(\tau)$ and the rate matrix $Q^c(\tau)$ between the dominant conformations. PCCA+ works with a projection of the matrix via a set of basis functions $\chi_i$. In discrete case, $\chi_i$ is a vector that gives information about the membership grade of each discretization cell to this dominant conformation $i$. So, is the cell $j$ 100% in the dominant conformation $i$, so is $\chi_{ij}=1$.

What is very advantageous about this projection of the operator is that the propagation and the projection commutes. This means, one obtain the same result for $(K^c(\tau))^k$(first compute $K(\tau)$, then iterate $k$ times) or $K^c(k\tau)$ (first iterate $k$ times and obtain $K(k\tau)$, then project) . 


Note that the commutation relation of the PCCA+ can be applied for the computation of the projected rate matrix $Q^c$ as well. 

**How do we compute the projected infinitesimal generator?**

In principle, there are two procedures that yields to the PCCA+-projected rate-matrix:
    1. project the Koopman matrix $K(\tau)$ with PCCA+, apply finite-differences/ matrix logarithm/Newton; 
    2. apply finite-differences/ matrix logarithm/Newton, then project the rate matrix into a smaller conformational space with PCCA+ 


_In the following, you can try on our own the two options and convince yourself that it works!_  
Here the application to the sequential decay dataset.

%with a difference compute with the operator norm $\Vert Q_{fd1}-Q_{fd2} \Vert= 0.0005$ (relative $\Vert Q_{fd1}-Q_{fd2} \Vert/\Vert Q_{fd1}\Vert=is 0.016$). So, the computation of the rate matrix will be not affected by the choice of one of the two options. 

In [1]:
import numpy as np 
from scipy.linalg import pinv
import cmdtools 
#add files for the computation
# sequential decay

In [10]:
#load transition matrix for a discretization of in 50 Voronoi cells of the spectrum
K = np.loadtxt("K_tutorial_50vor.txt")
#number of dominant conformations 
nclus = 3

In [7]:
#option 1
chi1 =  cmdtools.analysis.pcca.pcca(K,nclus) #membership functions for Q2
K_c =  pinv(chi1).dot((K.dot(chi1))) #PCCA+ projected transition matrix, (nclus x nclus)
Q_c1 = K_c - np.eye(nclus) #PCCA+ projected rate matrix
print(Q_c1)

[[-0.0068202   0.00795601 -0.00113582]
 [ 0.01255214 -0.02300592  0.01045377]
 [-0.00139596  0.01217041 -0.01077445]]


In [8]:
#option 2
Q2 = K - np.eye(50) # rate matrix 50x50
chi2 =  cmdtools.analysis.pcca.pcca(Q2,nclus) #membership functions for Q2
Q_c2 =  pinv(chi2).dot((K-np.eye(50)).dot(chi2)) #PCCA+ projected rate matrix, (nclus x nclus)
print(Q_c2)

[[-0.0068202   0.00795601 -0.00113582]
 [ 0.01255214 -0.02300592  0.01045377]
 [-0.00139596  0.01217041 -0.01077445]]


In [11]:
Q_c1 - Q_c2

array([[-5.20417043e-17,  5.39499001e-16,  2.54353830e-16],
       [-2.22044605e-16,  5.58580959e-16,  1.73472348e-17],
       [ 3.09214460e-16, -5.03069808e-17, -3.81639165e-17]])

 For the first option, one obtains \begin{align}
    Q_{c1} = \begin{pmatrix}
    -0.007&	0.008&	-0.001\\
0.013&	-0.023&	0.011\\
-0.001&	0.012&	-0.011
    \end{pmatrix};
\end{align} for the second option: \begin{align}
    Q_{c2} = \begin{pmatrix}
    -0.007&	0.008&	-0.001 \\
0.013&	-0.023&	0.010\\
-0.001&	0.012&	-0.011
    \end{pmatrix}
\end{align}


So we checked that the commutator works! Literature about it can be found i.e. in [M.Weber, A Subspace Approach to Molecular Markov State Models via a New Infinitesimal Generator, 2011](https://opus4.kobv.de/opus4-zib/frontdoor/index/index/docId/1402 "Theorem 2, Chap.3.3")
